cli: ensure --check flag works for piped-in code

not-an-aardvark opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

Previously, the --check CLI flag had no effect when run on code piped
from stdin. This commit updates the bootstrap logic to handle the
--check flag the same way regardless of whether the code is piped from
stdin.

Fixes: https://github.com/nodejs/node/issues/11680

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

lib

cjihrig commented on lib/internal/bootstrap_node.js:
```diff
@@ -127,18 +127,11 @@
-          // wrap it
-          source = Module.wrap(source);
-          // compile the script, this will throw if it fails
-          new vm.Script(source, {filename: filename, displayErrors: true});
```
Why drop the `filename`?

not-an-aardvark replied:
If I'm not mistaken, the filename doesn't affect whether a script is syntactically valid, and piped-in code has no filename.

cjihrig replied:
I think the filename gets displayed if there is an error, which might be helpful. Dropping that information would probably be a semver major. The piped-in scenario could just use `[stdin]`.

not-an-aardvark replied:
Ok, I'll add it back.

addaleax commented:
> Fixes: #11680

But only the first part of it, if I’m reading the code here correctly? I think we should support `-c -e` as well (or error out when those arguments are passed).

not-an-aardvark commented:
Correct, this currently does not fix `-c -e`. What do you think should be the behavior with `-c -e`? It seems like flags for "check syntax only" and "evaluate this code" would contradict each other, so I'm not sure about the correct behavior.

gibfahn commented:
>(or error out when those arguments are passed).

>What do you think should be the behavior with `-c -e`?

+1 to erroring out (with a suitable message).

not-an-aardvark commented:
I added a commit to throw when `-c` and `-e` are used together.

not-an-aardvark commented on lib/internal/bootstrap_node.js:
```diff
@@ -114,6 +114,10 @@
       if (process._eval != null && !process._forceRepl) {
+        if (process._syntax_check_only != null) {
+          console.error('--check and --eval flags are mutually exclusive.');
+          process.exit(1);
```
I couldn't find any other logic like this for flags, so I sort of improvised with the error message. Let me know if there's a better way to do this.

richardlau replied:
The exit code should be `9`, see https://nodejs.org/dist/latest-v7.x/docs/api/process.html#process_exit_codes.

https://github.com/nodejs/node/pull/12087 is doing something similar for the mutually exclusive `--use-bundled-ca` and `--use-openssl-ca` options.

cjihrig commented on test/parallel/test-cli-syntax.js:
```diff
@@ -82,3 +85,50 @@ const syntaxArgs = [
+  assert.strictEqual(c.status, 0, 'code == ' + c.status);
+});
+
+// should should throw if code piped from stdin with --check has bad syntax
```
Duplicate "should"

not-an-aardvark commented:
Updated to address comments.

CI: https://ci.nodejs.org/job/node-test-pull-request/7128/

jasnell commented:
Quick note: if the intention is to have this land for 8.0.0, it needs to be completed by April 4th.

jasnell commented on lib/internal/bootstrap_node.js:
```diff
@@ -114,6 +114,10 @@
+        if (process._syntax_check_only != null) {
+          console.error('--check and --eval flags are mutually exclusive.');
+          process.exit(9);
+        }
```
Should these be checked here or should they be checked in `node.cc`? Note that there are other mutually exclusive flags and there is an open PR that does this kind of check directly within `node.cc`

jasnell commented on test/parallel/test-cli-syntax.js:
```diff
@@ -82,3 +85,50 @@ const syntaxArgs = [
+      '--check and --eval flags are mutually exclusive.\n'
+    );
+
+    assert.strictEqual(c.status, 9, 'code == ' + c.status);
```
the assert message should likely read `===` since we're testing strictEqual

not-an-aardvark commented:
Updated, new CI: https://ci.nodejs.org/job/node-test-pull-request/7147/

@jasnell I couldn't find the other mutually exclusive flags that you were referring to, but I moved the check to `node.cc`. Let me know if there's another place in `node.cc` that it should go instead.

not-an-aardvark commented:
Actually, I see that another mutually-exclusive-flag check landed on master and is now causing a conflict with this PR. I suppose I put the check in the right place then. :smiley:

not-an-aardvark commented:
Rebased, new CI: https://ci.nodejs.org/job/node-test-pull-request/7148/

addaleax commented on test/parallel/test-cli-syntax.js:
```diff
@@ -82,3 +85,50 @@ const syntaxArgs = [
+  const match = c.stderr.match(/^SyntaxError: Unexpected identifier$/m);
+  assert(match, 'stderr incorrect');
+
+  assert.strictEqual(c.status, 1, 'code == ' + c.status);
```
All `==`s here should probably be `===`, but that can be fixed while landing

not-an-aardvark commented:
Landed in 3209a8ebf3aa2587c68c4b72e6c3b93176090807 and a5f91ab230c574d561780b6867d00f06fcc1e4de.

bzoz commented:
Windows CI is not green, `test-cli-syntax` fails on Windows (https://ci.nodejs.org/job/node-test-binary-windows/7416/):

```
not ok 32 parallel/test-cli-syntax
  ---
  duration_ms: 3.104
  severity: fail
  stack: |-
    
    assert.js:81
      throw new assert.AssertionError({
      ^
    AssertionError: 'c:\\workspace\\node-test-binary-windows\\RUN_SUBSET\\1\\VS_VERSION\\vcbt2015\\label\\win10\\Release\\node.exe: either --check o === 'c:\\workspace\\node-test-binary-windows\\RUN_SUBSET\\1\\VS_VERSION\\vcbt2015\\label\\win10\\Release\\node.exe: either --check o
        at c:\workspace\node-test-binary-windows\RUN_SUBSET\1\VS_VERSION\vcbt2015\label\win10\test\parallel\test-cli-syntax.js:127:12
        at Array.forEach (native)
        at c:\workspace\node-test-binary-windows\RUN_SUBSET\1\VS_VERSION\vcbt2015\label\win10\test\parallel\test-cli-syntax.js:123:20
        at Array.forEach (native)
        at Object.<anonymous> (c:\workspace\node-test-binary-windows\RUN_SUBSET\1\VS_VERSION\vcbt2015\label\win10\test\parallel\test-cli-syntax.js:122:19)
        at Module._compile (module.js:607:30)
        at Object.Module._extensions..js (module.js:618:10)
        at Module.load (module.js:516:32)
        at tryModuleLoad (module.js:466:12)
        at Function.Module._load (module.js:458:3)
  ...
```

not-an-aardvark commented:
Sorry about that. I'll be able to create a PR to fix it in about 2 hours if no one gets around to it before then.

Fishrock123 commented:
Not 100% sure I agree, it could be nice to pass source code directly to `-c` I guess.

I doubt anyone would use it though, so really... ¯\\\_(ツ)\_/¯

not-an-aardvark commented:
@Fishrock123 To clarify, it's possible to do `echo foo | node -c` (added in this PR). It's just that `node -c -e foo` is disallowed.
